prov/psm2: Add scalable endpoint support

j-xiong opened the pull request:
This set of commits implement scalable endpoints in the psm2 provider. It allows applications to utilize multiple hardware Tx/Rx contexts to achieve higher bandwidth over a single endpoint.

j-xiong commented:
@shefty The changes are big, but all confined within the psm2 provider.

shefty commented:
@j-xiong - Do you want me to do a detailed review of the changes?

j-xiong commented:
@shefty Any level of review is good. Depending on your time.

shefty commented on prov/psm2/src/psmx2.h:
```diff
@@ -723,6 +738,27 @@ struct psmx2_fid_ep {
+	uint8_t			id;
+	uint8_t			enabled;
+	size_t			ctxt_cnt;
+	struct psmx2_sep_ctxt	ctxts[0]; /* must be last element */
```
I think the standard is to use [ ] without the 0.

shefty commented on prov/psm2/src/psmx2.h:
```diff
@@ -690,18 +709,31 @@ struct psmx2_fid_cntr {
+
+struct psmx2_sep_addr {
+	int			ctxt_cnt;
+	struct psmx2_ctxt_addr	ctxt_addrs[0];
```
Can use [ ]

shefty commented on prov/psm2/src/psmx2_av.c:
```diff
@@ -32,6 +32,137 @@
+static void psmx2_am_sep_completion(void *buf)
+{
+	if (buf)
+		free(buf);
```
free can take NULL as input

shefty commented on prov/psm2/src/psmx2_av.c:
```diff
@@ -32,6 +32,137 @@
+			n = sep->ctxt_cnt;
+			buflen = n * sizeof(psm2_epid_t);
+			if (n) {
+				buf = malloc(buflen);
```
need to check for malloc failure

shefty commented on prov/psm2/src/psmx2_av.c:
```diff
@@ -32,6 +32,137 @@
+			p->ctxt_cnt = n;
+			for (j=0; j<n; j++) {
+				p->ctxt_addrs[j].epid = ((psm2_epid_t *)src)[j];
+				p->ctxt_addrs[j].epaddr = (psm2_epaddr_t)FI_ADDR_NOTAVAIL;
```
p may be NULL if calloc fails

shefty commented on prov/psm2/src/psmx2_av.c:
```diff
@@ -223,38 +223,50 @@ int psmx2_epid_to_epaddr(struct psmx2_fid_domain *domain,
+			FI_WARN(&psmx2_prov, FI_LOG_AV,
+				"fatal error: unable to translate epid %lx to epaddr.\n",
+				av->sepaddrs[idx]->ctxt_addrs[ctxt].epid);
+			exit(-1);
```
calling exit from a library is rather rude :) Maybe assert() and return NULL?

shefty commented on prov/psm2/src/psmx2_tagged.c:
```diff
@@ -996,7 +1065,10 @@ ssize_t psmx2_tagged_sendv_generic(struct fid_ep *ep,
+	if (av && PSMX2_SEP_ADDR_TEST(dest_addr)) {
+		psm2_epaddr = psmx2_av_translate_sep(av, ep_priv->trx_ctxt, dest_addr);
+		vlane = 0;
+	} else  if (av && av->type == FI_AV_TABLE) {
```
Is there some way to encapsulate these changes into a call?  This same set of changes is repeated about a dozen times or more.

j-xiong replied:
I think it would be better handled via a separate PR. The consolidation will be applied not only to the newly added code, but also the existing code in the other conditional branches.

shefty commented on prov/psm2/src/psmx2_init.c:
```diff
@@ -62,6 +65,23 @@ static void psmx2_init_env(void)
+
+	if ((*func)(mask) != mask)
+		return 0;
+
```
Need dlclose

shefty commented on prov/psm2/src/psmx2_init.c:
```diff
@@ -122,6 +120,53 @@ static int psmx2_init_lib(void)
+	close(fd);
+
+	return ret;
+}
```
I think you can use fi_read_file() -- see fi_file.h

shefty commented:
I reviewed this by examining each commit individually, so not all comments apply to the final set of changes.

j-xiong commented:
@shefty Thanks! Will address the issues and update the PR.

j-xiong commented:
@shefty The commits are updated. I have addressed all the comments except for the SEP address testing (will optimize in a future PR) and the missing of dlclose.

shefty commented:
👍 thanks

rzambre commented:
When I to use it with opa-psm2, my `domain->ops->scalable_ep` gets initialized with `fi_no_scalable_ep`. That's because `psmx2_sep_ok` returns 0 if `PSM2_MULTI_EP_CAP` is not defined. But I don't see `PSM2_MULTI_EP_CAP` in the `01org/opa-psm2` repo.

1) Will this flag be defined in a future release of PSM2?
2) Is there any way to use this feature with `01org/opa-psm2`? Hacking out the `#ifdef` won't work because `psm2_get_capability_mask`, `psm2_ep_epid_lookup2` haven't made it upstream to opa-psm2.

j-xiong commented:
1. yes
2. you have to wait until the new feature gets upstream.

I agree this can be confusing but PSM2 and libfabric have different development cycle. We have to get the provider update before PSM2 in order to fit into the 1.5 release time frame. Although it has been tested internally, right now please consider the scalable endpoint support experimental.

rzambre commented:
Thank you! Any estimate on when the new feature will make it upstream into `opa-psm2`?

j-xiong commented:
I don't have the information.

rzambre commented:
Could you point me to an example-case that uses this? Or a development branch of the newer psm2? We are running some experiments here and so, we don't need the production version of the newer PSM2.

j-xiong commented:
Here is an example: https://github.com/j-xiong/ofi-test.

Unfortunately I can't give you a development branch of the newer psm2. You have to wait for that being pushed to github.
